frontend/src/lib/utils.ts

```diff
@@ -100,23 +88,18 @@ export function requireNonNull<T>(obj: T | null | undefined): T {
 }
 
 /**
- * Provide an ImmutableSet of SimpleElements by walking a BlockElement to
- * its leaves.
+ * A type predicate that is true if the given value is not undefined.
  */
-export function flattenElements(
-  elements: BlockElement
-): ImmutableSet<SimpleElement> {
-  return elements.reduce(
-    (flattened: ImmutableSet<SimpleElement>, reportElement: ReportElement) => {
-      const element = reportElement.get("element")
-
-      if (element instanceof List) {
-        return flattened.union(flattenElements(element as BlockElement))
-      }
-      return flattened.union(ImmutableSet.of(element as SimpleElement))
-    },
-    ImmutableSet.of<SimpleElement>()
-  )
+export function notUndefined<T>(value: T | undefined): value is T {
```

Comment:
nit: not sure that this is justified having its own util function, since the logic is more easily examined directly wherever it's referenced, plus we're almost guaranteed to inconsistently use it.

Reply:
I softly disagree with this, for similar reasons to `notNull` above - type guards have syntax that's tricky to remember (`value is T` is a strange-looking return type), but that's very easy to understand when you see it, IMO.

In both cases, these are functions that allow for more fluent reading of code. For example, in `App.handleReportFinished`, we're using it to filter an Array:

```typescript
const activeWidgetIds = new Set(
  Array.from(this.state.elements.getElements())
    .map(element => getElementWidgetID(element))
    .filter(notUndefined)
)
```

The alternative is:

```typescript
const activeWidgetIds = new Set(
  Array.from(this.state.elements.getElements())
    .map(element => getElementWidgetID(element))
    .filter((value): value is string => value !== undefined)
)
```

Which just doesn't scan at all, IMO.

Reply:
Ah, I wasn't aware of the "value is T" requirement -- I think that's a reasonable argument for keeping it as-is. Thanks for the explanation!